File: test/core/services/recents_repository_test.dart
import 'package:aruku/core/models/geo_point.dart';
import 'package:aruku/core/models/recent_place.dart';
import 'package:aruku/core/services/recents_repository.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    SharedPreferences.setMockInitialValues({});
  });

  Future<RecentsRepository> makeRepo() async {
    final prefs = await SharedPreferences.getInstance();
    return RecentsRepository(prefs);
  }

  Future<RecentsRepository> makeOriginsRepo() async {
    final prefs = await SharedPreferences.getInstance();
    return RecentsRepository(prefs, key: RecentsRepository.originsKey);
  }

  RecentPlace dest(
    String name, {
    String? placeId,
    GeoPoint? latLng,
    DateTime? usedAt,
  }) {
    return RecentPlace(
      name: name,
      placeId: placeId,
      latLng: latLng,
      usedAt: usedAt,
    );
  }

  test('保存ゼロ件なら load は空リスト', () async {
    final repo = await makeRepo();
    expect(await repo.load(), isEmpty);
  });

  test('add したものが load で先頭から取り出せる', () async {
    final repo = await makeRepo();
    await repo.add(dest('東京駅', placeId: 'p1'));
    await repo.add(dest('渋谷駅', placeId: 'p2'));
    final list = await repo.load();
    expect(list.map((e) => e.name).toList(), ['渋谷駅', '東京駅']);
  });

  test('同じ placeId は重複排除され、最新が先頭に来る', () async {
    final repo = await makeRepo();
    await repo.add(dest('東京駅', placeId: 'p1'));
    await repo.add(dest('渋谷駅', placeId: 'p2'));
    await repo.add(dest('東京駅', placeId: 'p1'));
    final list = await repo.load();
    expect(list.map((e) => e.placeId).toList(), ['p1', 'p2']);
  });

  test('placeId が無い場合は name で重複排除', () async {
    final repo = await makeRepo();
    await repo.add(dest('カフェ'));
    await repo.add(dest('公園'));
    await repo.add(dest('カフェ'));
    final list = await repo.load();
    expect(list.map((e) => e.name).toList(), ['カフェ', '公園']);
  });

  test('最大件数を超えると古いものから捨てる', () async {
    final repo = await makeRepo();
    for (var i = 0; i < RecentsRepository.maxItems + 3; i++) {
      await repo.add(dest('place_$i', placeId: 'id_$i'));
    }
    final list = await repo.load();
    expect(list.length, RecentsRepository.maxItems);
    expect(list.first.placeId, 'id_${RecentsRepository.maxItems + 2}');
  });

  test('clear で空になる', () async {
    final repo = await makeRepo();
    await repo.add(dest('東京駅', placeId: 'p1'));
    await repo.clear();
    expect(await repo.load(), isEmpty);
  });

  test('出発地用キーは目的地履歴と混ざらず独立して保存される', () async {
    final destinations = await makeRepo();
    final origins = await makeOriginsRepo();

    await destinations.add(dest('東京駅', placeId: 'd1'));
    await origins.add(dest('自宅', placeId: 'o1'));

    expect((await destinations.load()).map((e) => e.name).toList(), ['東京駅']);
    expect((await origins.load()).map((e) => e.name).toList(), ['自宅']);
  });

  test('出発地リポジトリは destinations キーを書き換えない', () async {
    final destinations = await makeRepo();
    final origins = await makeOriginsRepo();

    await destinations.add(dest('東京駅', placeId: 'd1'));
    await origins.clear();

    expect((await destinations.load()).map((e) => e.name).toList(), ['東京駅']);
  });

  test('並行 add でも書き込みが失われず全件保存される', () async {
    final repo = await makeRepo();
    await Future.wait([
      for (var i = 0; i < 5; i++) repo.add(dest('place_$i', placeId: 'id_$i')),
    ]);
    final list = await repo.load();
    expect(list.length, 5);
    expect(list.map((e) => e.placeId).toSet(), {
      'id_0',
      'id_1',
      'id_2',
      'id_3',
      'id_4',
    });
  });
}

File: apps/web/test/places/recents-repository.test.ts
// 移植元: lib/core/services/recents_repository.dart と
// test/core/services/recents_repository_test.dart
//
// 書き込みの直列化（_writeLock）は移植していない。localStorage が同期なので
// load→変更→save の間に別の操作が割り込む余地が無い。その反証が
// 「連続して add しても取りこぼさない」。

import { describe, expect, it, vi } from 'vitest';

import { GeoPoint } from '@aruku/engine/models/geo-point';

import {
  createRecentsRepository,
  destinationsKey,
  maxRecents,
  originsKey,
  type KeyValueStore,
} from '../../src/places/recents-repository';
import type { RecentPlace } from '../../src/places/recent-place';

function memoryStore(initial: Record<string, string> = {}): KeyValueStore {
  const map = new Map(Object.entries(initial));
  return {
    getItem: (key) => map.get(key) ?? null,
    setItem: (key, value) => void map.set(key, value),
    removeItem: (key) => void map.delete(key),
  };
}

function place(name: string, overrides: Partial<RecentPlace> = {}): RecentPlace {
  return {
    name,
    placeId: null,
    latLng: new GeoPoint(35.6, 139.7),
    address: null,
    usedAt: null,
    ...overrides,
  };
}

describe('読み込み', () => {
  it('何も無ければ空', () => {
    expect(createRecentsRepository(memoryStore()).load()).toEqual([]);
  });

  it('空文字は空として読む', () => {
    const repo = createRecentsRepository(memoryStore({ [destinationsKey]: '' }));

    expect(repo.load()).toEqual([]);
  });

  it('壊れた JSON は空として読む', () => {
    const repo = createRecentsRepository(
      memoryStore({ [destinationsKey]: '{壊れている' }),
    );

    expect(repo.load()).toEqual([]);
  });

  it('配列でない JSON は空として読む', () => {
    const repo = createRecentsRepository(
      memoryStore({ [destinationsKey]: '{"name":"渋谷"}' }),
    );

    expect(repo.load()).toEqual([]);
  });

  // 1件の壊れで履歴を全部落とさない。読めたものは読む。
  it('地点として成立しない要素だけ読み飛ばす', () => {
    const repo = createRecentsRepository(
      memoryStore({
        [destinationsKey]: JSON.stringify([
          { name: '渋谷駅' },
          { placeId: 'p1' },
          'ごみ',
          { name: '新宿駅' },
        ]),
      }),
    );

    expect(repo.load().map((p) => p.name)).toEqual(['渋谷駅', '新宿駅']);
  });
});

describe('追加', () => {
  it('先頭へ積む', () => {
    const repo = createRecentsRepository(memoryStore());

    repo.add(place('渋谷駅'));
    repo.add(place('新宿駅'));

    expect(repo.load().map((p) => p.name)).toEqual(['新宿駅', '渋谷駅']);
  });

  it('同じ地点は重複させず先頭へ繰り上げる', () => {
    const repo = createRecentsRepository(memoryStore());

    repo.add(place('渋谷駅', { placeId: 'p1' }));
    repo.add(place('新宿駅', { placeId: 'p2' }));
    repo.add(place('渋谷駅', { placeId: 'p1' }));

    expect(repo.load().map((p) => p.name)).toEqual(['渋谷駅', '新宿駅']);
  });

  it('上限を超えたぶんは古い方から落とす', () => {
    const repo = createRecentsRepository(memoryStore());

    for (let i = 0; i < maxRecents + 3; i++) repo.add(place(`駅${i}`));

    const names = repo.load().map((p) => p.name);
    expect(names).toHaveLength(maxRecents);
    expect(names[0]).toBe(`駅${maxRecents + 2}`);
    expect(names).not.toContain('駅0');
  });

  it('使った時刻を打つ', () => {
    const now = new Date('2026-09-13T01:00:00.000Z');
    const repo = createRecentsRepository(memoryStore(), destinationsKey, () => now);

    repo.add(place('渋谷駅'));

    expect(repo.load()[0]!.usedAt).toEqual(now);
  });

  it('すでに時刻を持つ記録は打ち直さない', () => {
    const stamped = new Date('2026-01-01T00:00:00.000Z');
    const repo = createRecentsRepository(
      memoryStore(),
      destinationsKey,
      () => new Date('2026-09-13T01:00:00.000Z'),
    );

    repo.add(place('渋谷駅', { usedAt: stamped }));

    expect(repo.load()[0]!.usedAt).toEqual(stamped);
  });

  // localStorage は同期なので、Dart 版の _writeLock が防いでいた
  // 「load→save の間に別の add が割り込んで互いを上書きする」が起き得ない。
  it('連続して add しても取りこぼさない', () => {
    const repo = createRecentsRepository(memoryStore());

    repo.add(place('A'));
    repo.add(place('B'));
    repo.add(place('C'));

    expect(repo.load().map((p) => p.name)).toEqual(['C', 'B', 'A']);
  });

  it('座標と住所を保って往復する', () => {
    const repo = createRecentsRepository(memoryStore());

    repo.add(place('渋谷駅', { placeId: 'p1', address: '東京都渋谷区' }));

    const [restored] = repo.load();
    expect(restored!.latLng).toEqual(new GeoPoint(35.6, 139.7));
    expect(restored!.address).toBe('東京都渋谷区');
    expect(restored!.placeId).toBe('p1');
  });
});

describe('消去', () => {
  it('履歴を空にする', () => {
    const repo = createRecentsRepository(memoryStore());
    repo.add(place('渋谷駅'));

    repo.clear();

    expect(repo.load()).toEqual([]);
  });
});

describe('系統の独立', () => {
  it('目的地と出発地は互いに混ざらない', () => {
    const store = memoryStore();
    const destinations = createRecentsRepository(store, destinationsKey);
    const origins = createRecentsRepository(store, originsKey);

    destinations.add(place('渋谷駅'));
    origins.add(place('自宅'));

    expect(destinations.load().map((p) => p.name)).toEqual(['渋谷駅']);
    expect(origins.load().map((p) => p.name)).toEqual(['自宅']);
  });
});

// Safari のプライベートモード、サイトデータを止めた設定、容量超過。どれも
// localStorage のアクセス**自体**が投げる。履歴は利便のための機能なので、
// 読めない・書けないことで検索そのものを落としてはいけない。
describe('localStorage が使えない環境', () => {
  it('読めなければ空として扱う', () => {
    const repo = createRecentsRepository({
      getItem: () => {
        throw new Error('SecurityError');
      },
      setItem: () => {},
      removeItem: () => {},
    });

    expect(repo.load()).toEqual([]);
  });

  it('書けなくても落ちない', () => {
    const setItem = vi.fn(() => {
      throw new Error('QuotaExceededError');
    });
    const repo = createRecentsRepository({
      getItem: () => null,
      setItem,
      removeItem: () => {},
    });

    expect(() => repo.add(place('渋谷駅'))).not.toThrow();
    expect(setItem).toHaveBeenCalled();
  });

  it('消せなくても落ちない', () => {
    const repo = createRecentsRepository({
      getItem: () => null,
      setItem: () => {},
      removeItem: () => {
        throw new Error('SecurityError');
      },
    });

    expect(() => repo.clear()).not.toThrow();
  });
});
